File: shared/src/commonMain/kotlin/com/jetbrains/spacetutorial/texaswatch/ui/OffenderCard.kt
package com.jetbrains.spacetutorial.texaswatch.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import coil3.compose.SubcomposeAsyncImage
import coil3.compose.LocalPlatformContext
import coil3.request.ImageRequest
import coil3.request.crossfade
import com.jetbrains.spacetutorial.texaswatch.entity.OffenderSummary
import com.jetbrains.spacetutorial.texaswatch.theme.TexasWatchTheme

// ── Offender Card ─────────────────────────────────────────────────────────────
// Same layout as SpeakerCard: 96dp photo on the left, name + subtitle on right

@Composable
fun OffenderCard(
    offender: OffenderSummary,
    onClick: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .clip(TexasWatchTheme.shapes.medium)
            .then(if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier)
            .background(TexasWatchTheme.colors.cardBackground)
            .padding(12.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        OffenderPhoto(
            photoUrl = offender.photoUrl,
            initials = initials(offender.firstName, offender.lastName),
            modifier = Modifier.size(96.dp),
        )
        Column {
            Text(
                text = offender.fullName,
                style = TexasWatchTheme.typography.h3,
                color = TexasWatchTheme.colors.primaryText,
            )
            Spacer(Modifier.size(6.dp))
            Text(
                text = offenderSubtitle(offender),
                style = TexasWatchTheme.typography.text2,
                color = TexasWatchTheme.colors.secondaryText,
            )
            Spacer(Modifier.size(4.dp))
            if (!offender.address.isNullOrBlank()) {
                Text(
                    text = offender.address.orEmpty(),
                    style = TexasWatchTheme.typography.text2,
                    color = TexasWatchTheme.colors.secondaryText,
                )
            }
        }
    }
}

// ── Offender Photo ────────────────────────────────────────────────────────────
// Uses Coil SubcomposeAsyncImage so we can show InitialsAvatar in loading/error states

@Composable
fun OffenderPhoto(
    photoUrl: String?,
    initials: String,
    modifier: Modifier = Modifier,
) {
    val clippedModifier = modifier
        .clip(TexasWatchTheme.shapes.medium)
        .background(TexasWatchTheme.colors.surfaceBackground)

    if (!photoUrl.isNullOrBlank()) {
        SubcomposeAsyncImage(
            model = ImageRequest.Builder(LocalPlatformContext.current)
                .data(photoUrl)
                .crossfade(true)
                .build(),
            contentDescription = null,
            modifier = clippedModifier,
            contentScale = ContentScale.Crop,
            loading = { InitialsAvatar(initials = initials, modifier = Modifier.fillMaxSize()) },
            error = { InitialsAvatar(initials = initials, modifier = Modifier.fillMaxSize()) },
        )
    } else {
        InitialsAvatar(initials = initials, modifier = clippedModifier)
    }
}

// ── Initials fallback avatar ──────────────────────────────────────────────────

@Composable
fun InitialsAvatar(initials: String, modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .clip(TexasWatchTheme.shapes.medium)
            .background(TexasWatchTheme.colors.primaryAccent),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = initials,
            style = TexasWatchTheme.typography.h2,
            color = TexasWatchTheme.colors.invertedText,
            fontWeight = FontWeight.Bold,
        )
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

private fun initials(firstName: String, lastName: String): String {
    val f = firstName.trim().firstOrNull()?.uppercaseChar() ?: ""
    val l = lastName.trim().firstOrNull()?.uppercaseChar() ?: ""
    return "$f$l"
}

private fun offenderSubtitle(offender: OffenderSummary): String {
    val parts = mutableListOf<String>()
    if (offender.age != null) parts.add("Age ${offender.age}")
    parts.add("DPS: ${offender.dpsNumber}")
    return parts.joinToString("  ·  ")
}

File: iosApp/iosApp/OffenderCard.swift
import SwiftUI
import Shared

// ── Native iOS Offender Card ──────────────────────────────────────────────────

struct OffenderCard: View {
    let offender: OffenderSummary
    @Environment(\.twColors) private var colors
    @Environment(\.twTypography) private var typography

    var body: some View {
        HStack(spacing: 12) {
            OffenderPhoto(photoUrl: offender.photoUrl, initials: initials)
                .frame(width: 96, height: 96)
                .clipShape(RoundedRectangle(cornerRadius: 12))

            VStack(alignment: .leading, spacing: 4) {
                Text(offender.fullName)
                    .font(typography.h3)
                    .foregroundStyle(colors.primaryText)

                Text(subtitle)
                    .font(typography.text2)
                    .foregroundStyle(colors.secondaryText)

                if let address = offender.address, !address.isEmpty {
                    Text(address)
                        .font(typography.text2)
                        .foregroundStyle(colors.secondaryText)
                        .lineLimit(2)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(12)
        .background(colors.mainBackground)
        .clipShape(RoundedRectangle(cornerRadius: 0))
    }

    private var initials: String {
        let f = offender.firstName.first.map { String($0).uppercased() } ?? ""
        let l = offender.lastName.first.map { String($0).uppercased() } ?? ""
        return f + l
    }

    private var subtitle: String {
        var parts: [String] = []
        if let age = offender.age { parts.append("Age \(age)") }
        parts.append("DPS: \(offender.dpsNumber)")
        return parts.joined(separator: "  ·  ")
    }
}

// ── Photo with async loading + initials fallback ──────────────────────────────

struct OffenderPhoto: View {
    let photoUrl: String?
    let initials: String

    var body: some View {
        if let url = photoUrl.flatMap(URL.init), !url.absoluteString.isEmpty {
            AsyncImage(url: url) { phase in
                switch phase {
                case .success(let image):
                    image.resizable().scaledToFill()
                case .failure, .empty:
                    InitialsAvatar(initials: initials)
                @unknown default:
                    InitialsAvatar(initials: initials)
                }
            }
        } else {
            InitialsAvatar(initials: initials)
        }
    }
}

// ── Initials fallback ─────────────────────────────────────────────────────────

struct InitialsAvatar: View {
    let initials: String
    @Environment(\.twColors) private var colors

    var body: some View {
        ZStack {
            colors.primaryAccent
            Text(initials)
                .font(.title2.bold())
                .foregroundStyle(colors.invertedText)
        }
    }
}
